src/librustc_mir/transform/const_prop.rs

```diff
@@ -988,8 +1023,61 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             | TerminatorKind::GeneratorDrop
             | TerminatorKind::FalseEdges { .. }
             | TerminatorKind::FalseUnwind { .. } => {}
-            //FIXME(wesleywiser) Call does have Operands that could be const-propagated
-            TerminatorKind::Call { .. } => {}
+            TerminatorKind::Call { ref mut args, .. } => {
+                // FIXME(felix91gr) Finish this
+                for opr in args {
+                    // If it isn't a place, then it's a constant,
+                    // and thus we don't need to propagate towards it.
+                    trace!("Checking if arg {:?} is not a constant", &opr);
+                    if let Some(place) = opr.place() {
+                        // No constants from here on
+                        trace!("Obtained place: {:?}", &place);
+                        trace!("Checking if arg {:?} has no projections", &opr);
+                        if let Some(local) = place.as_local() {
+                            // No projections from here on
+                            trace!("Place has no projections. Local is {:?}", &local);
+                            let place_evaluated = self.eval_place(place);
+                            trace!("Place is evaluated  as {:?}", place_evaluated);
+                            if let Some(op_ty) = place_evaluated {
+                                trace!("OpTy of the place: {:?}", &op_ty);
+                                use std::ops::Deref;
+                                if let interpret::Operand::Immediate(immediate) = op_ty.deref() {
```

Comment:
This down to 1049 could be replaced with `self.ecx.read_scalar()`

Reply:
That almost works, but I get this:

`error: internal compiler error: src/librustc_mir/interpret/operand.rs:69: Got a wide pointer where a scalar was expected`

And taking a backtrace, here's where it comes from:

<details>

```
thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:904:9
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::sys_common::backtrace::print
   4: std::panicking::default_hook::{{closure}}
   5: std::panicking::default_hook
   6: rustc_driver::report_ice
   7: std::panicking::rust_panic_with_hook
   8: std::panicking::begin_panic
   9: rustc_errors::HandlerInner::bug
  10: rustc_errors::Handler::bug
  11: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
  12: rustc_middle::ty::context::tls::with_opt::{{closure}}
  13: rustc_middle::ty::context::tls::with_opt
  14: rustc_middle::util::bug::opt_span_bug_fmt
  15: rustc_middle::util::bug::bug_fmt
  16: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::read_scalar
  17: <rustc_mir::transform::const_prop::ConstPropagator as rustc_middle::mir::visit::MutVisitor>::visit_terminator
```
</details>

If we go to `read_scalar`, we see it calling `to_scalar_or_undef`:

```rust
    /// Read a scalar from a place
    pub fn read_scalar(
        &self,
        op: OpTy<'tcx, M::PointerTag>,
    ) -> InterpResult<'tcx, ScalarMaybeUndef<M::PointerTag>> {
        Ok(self.read_immediate(op)?.to_scalar_or_undef())
    }
```

Which in turn looks like this:

```rust
    #[inline]
    pub fn to_scalar_or_undef(self) -> ScalarMaybeUndef<Tag> {
        match self {
            Immediate::Scalar(val) => val,
            Immediate::ScalarPair(..) => bug!("Got a wide pointer where a scalar was expected"),
        }
    }
```

And so we need to filter the `ScalarPair`s or handle them somehow. I'm gonna try the first one.

But I like a lot how shorter this code is now :smile:

Reply:
Ah, yes. You also need to call `should_const_prop()` to see if we're allowed to make this change

https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/const_prop.rs#L740

That should be done in-between the calls to `self.eval_operand()` and `self.ecx.read_scalar()`.

Reply:
These might be some famous last words, but this change appears to be working. I suspect that the call to `should_const_prop` is the key piece that was missing (although everything else definitely helps make the code clearer)

Reply:
:)

Reply:
IT WORKED! :D